core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -135,14 +147,61 @@
    */
   @SafeVarargs
   public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
-    Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMap =
-        new HashMap<>(otherFields.length + 1);
-    indexMap.put(field, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
+    // count the numbers of two index types
+    int uniqIndexLenth = 0;
+    int dupIndexLenth = 0;
+
+    if (field.isUnique()) {
+      uniqIndexLenth = 1;
+    } else {
+      dupIndexLenth = 1;
+    }
+
     for (FieldIndex<T> fieldIndex : otherFields) {
-      indexMap.put(fieldIndex, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
+      if (fieldIndex.isUnique()) {
+        uniqIndexLenth++;
+      } else {
+        dupIndexLenth++;
+      }
+    }
+
+    // initialtion
+    Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMapDup = null;
+    Map<FieldIndex<T>, ConcurrentHashMap<Object, T>> indexMapUniq = null;
+    if (uniqIndexLenth > 0) {
+      indexMapUniq = new HashMap<>(uniqIndexLenth);
+    }
+    if (dupIndexLenth > 0) {
+      indexMapDup = new HashMap<>(dupIndexLenth);
+    }
+
+    if (field.isUnique()) {
+      indexMapUniq.put(field, new ConcurrentHashMap<Object, T>(8, 0.95f, 8));
+    } else {
+      indexMapDup.put(field, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
+    }
+
+    for (FieldIndex<T> fieldIndex : otherFields) {
+      if (fieldIndex.isUnique()) {
+        indexMapUniq.put(fieldIndex, new ConcurrentHashMap<Object, T>(8, 0.95f, 8));
+      } else {
+        indexMapDup.put(fieldIndex,
+            new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
+      }
     }
     // read only, so it is thread safe and allows concurrent access.
-    mIndexMap = Collections.unmodifiableMap(indexMap);
+    if (uniqIndexLenth > 0) {
+      mIndexMapUniq = Collections.unmodifiableMap(indexMapUniq);
+    } else {
```

Comment:
have you evaluated the overhead of always initializing the member variables to an empty map (as opposed to null)? this would increase readability because the code does not need to check if the maps are null